src/Desktop/ArcGISRuntime.Desktop.Samples/Samples/Map/AuthorMap/AuthorMap.xaml.cs

```diff
@@ -0,0 +1,497 @@
+        // resource is attempted
+        public async Task<Credential> CreateCredentialAsync(CredentialRequestInfo info)
+        {
+            OAuthTokenCredential credential = null;
+
+            try
+            {
+                // Create generate token options if necessary
+                if (info.GenerateTokenOptions == null)
+                {
+                    info.GenerateTokenOptions = new GenerateTokenOptions();
+                }
+
+                // IOAuthAuthorizeHandler will challenge the user for credentials
+                credential = await AuthenticationManager.Current.GenerateCredentialAsync
+                    (
+                            info.ServiceUri,
+                            info.GenerateTokenOptions
+                    ) as OAuthTokenCredential;
+            }
+            catch (Exception ex)
+            {
+                // Exception will be reported in calling function
+                throw (ex);
+            }
+
+            return credential;
+        }
+        #endregion
+    }
+
+    #region Helper class to display the OAuth authorization challenge
+    public class OAuthAuthorize : IOAuthAuthorizeHandler
+    {
+        // Window to contain the OAuth UI
+        private Window _window;
+        // Use a TaskCompletionSource to track the completion of the authorization
+        private TaskCompletionSource<IDictionary<string, string>> _tcs;
+        // URL for the authorization callback result (the redirect URI configured for your application)
+        private string _callbackUrl;
+        // URL that handles the OAuth request
+        private string _authorizeUrl;
+
+        // Function to handle authorization requests, takes the URIs for the secured service, the authorization endpoint, and the redirect URI
+        public Task<IDictionary<string, string>> AuthorizeAsync(Uri serviceUri, Uri authorizeUri, Uri callbackUri)
+        {
+            // If the TaskCompletionSource or Window are not null, authorization is in progress
+            if (_tcs != null || _window != null)
+            {
+                // Allow only one authorization process at a time
+                throw new Exception(); 
+            }
+
+            // Store the authorization and redirect URLs
+            _authorizeUrl = authorizeUri.AbsoluteUri;
+            _callbackUrl = callbackUri.AbsoluteUri;
+
+            // Create a task completion source
+            _tcs = new TaskCompletionSource<IDictionary<string, string>>();
+            var tcs = _tcs;
+
+            // Call a function to show the login controls, make sure it runs on the UI thread for this app
+            var dispatcher = Application.Current.Dispatcher;
+            if (dispatcher == null || dispatcher.CheckAccess())
+                AuthorizeOnUIThread(_authorizeUrl);
+            else
+            {
+                dispatcher.BeginInvoke((Action)(() => AuthorizeOnUIThread(_authorizeUrl)));
```

Comment:
I am definitely not en expert in delegates but can this be simplified into multiple lines of code? It is one heck of a complex statement.